hw/ip/otbn/util/make_mulqacc.py

```diff
@@ -0,0 +1,272 @@
+            acc_max += ((lhs_max * rhs_max) << shift_bits)
+            lines.append(
+                Line([], None, make_operand(lhs_regs, lhs_idx),
+                     make_operand(rhs_regs, rhs_idx), shift_bits))
+
+        if acc_max > mask(256):
+            raise ValueError(
+                f'Accumulator overflow while computing limb {i} of result!')
+
+        # Every second limb (except for the final word), write back a half-word
+        # of the result as part of the last instruction.
+        if not is_final_word and i % 2 == 1:
+            lines[-1].modifiers.append('so')
+            lines[-1].dst = make_half_word_writeback(result_regs, i)
+            # Update the accumulator bounds to reflect a half-word shift right.
+            before = int.bit_length(acc_max)
+            acc_max = mask(max(0, int.bit_length(acc_max) - (WDR_SIZE // 2)))
+            # print(i, before, int.bit_length(acc_max))
+
+    # It's possible that the last result term crosses a word boundary. If so,
+    # we need to add an instruction that just writes back the accumulator to
+    # the final word without modifying it. We assume here that w31 = 0.
+    if acc_max != 0:
+        lines.append(Line([], None, 'w31.0', 'w31.0', 0))
+
+    # Write back the final word of the result by adding a `.wo` modifier to the
+    # last instruction.
+    lines[-1].modifiers.append('wo')
+    lines[-1].dst = result_regs[result_nwords - 1]
+
+    # Ensure that the first instruction zeroes the accumulator.
+    lines[0].modifiers.append('z')
+
+    return lines
+
+
+def format_block(lines: List[Line]) -> List[str]:
+    '''Format the lines so that the left-hand operands are all aligned.'''
+    align = max([line.min_lhs_align() for line in lines])
+    return [line.pretty(align) for line in lines]
+
+
+def is_wdr(reg_name: str) -> bool:
+    if not reg_name.startswith('w'):
+        return False
+    if len(reg_name) < 1:
+        return False
+    reg_num = reg_name[1:]
+    if not reg_num.isnumeric():
+        return False
+    reg_idx = int(reg_num)
+    return (0 <= reg_idx) and (reg_idx <= 31)
+
+
+def check_regs(regs: List[str], bits: int) -> None:
+    '''Checks that register names are valid and there are enough registers to
+       store the given number of bits.'''
+    for reg_name in regs:
+        if not is_wdr(reg_name):
+            raise ValueError(f'{reg_name} is not a valid wide data register.')
+    if len(regs) * WDR_SIZE < bits:
+        raise ValueError(
+            f'Register list {regs} does not have enough space to store a '
+            f'{bits}-bit number.')
+    return
+
+
+def main() -> int:
+    parser = argparse.ArgumentParser(description=(
+        'Print out a bn.mulqacc block for the given bignum product.'))
+    parser.add_argument('--lhs-bits',
+                        type=int,
+                        required=True,
+                        help=('Number of bits in the left-hand operand.'))
```

Comment:
Nit: No need for the extra parens here: they're only really useful if you want to do something like
```
... = ('foo '
       'bar')
```